x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java

```diff
@@ -0,0 +1,402 @@
+        return version.onOrAfter(MINIMUM_COMPATIBLE);
+    }
+
+    public boolean after(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? false : true;
+        }
+        return version.id < id;
+    }
+
+    public boolean onOrAfter(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? false : true;
+        }
+        return version.id <= id;
+    }
+
+    public boolean before(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? true : false;
+        }
+        return version.id > id;
+    }
+
+    public boolean onOrBefore(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? true : false;
+        }
+        return version.id >= id;
+    }
+
+    public boolean between(MlConfigVersion lowerInclusive, MlConfigVersion upperExclusive) {
+        if (upperExclusive.onOrBefore(lowerInclusive)) throw new IllegalArgumentException();
+        return onOrAfter(lowerInclusive) && before(upperExclusive);
+    }
+
+    public static MlConfigVersion fromVersion(Version version) {
+        return fromId(version.id);
+    }
+
+    public static Version toVersion(MlConfigVersion mlConfigVersion) {
+        if (mlConfigVersion.id < MINIMUM_COMPATIBLE.id) {
+            throw new IllegalArgumentException("Cannot convert " + mlConfigVersion.id + ". Incompatible version");
+        }
+        return Version.fromId(mlConfigVersion.id);
+    }
+
+    public static MlConfigVersion getMinMlConfigVersion(DiscoveryNodes nodes, String mlConfigVersionNodeAttrName) {
+        return getMinMaxMlConfigVersion(nodes, mlConfigVersionNodeAttrName).v1();
+    }
+
+    public static MlConfigVersion getMaxMlConfigVersion(DiscoveryNodes nodes, String mlConfigVersionNodeAttrName) {
+        return getMinMaxMlConfigVersion(nodes, mlConfigVersionNodeAttrName).v2();
+    }
+
+    public static Tuple<MlConfigVersion, MlConfigVersion> getMinMaxMlConfigVersion(
+        DiscoveryNodes nodes,
+        String mlConfigVersionNodeAttrName
+    ) {
+        MlConfigVersion minMlConfigVersion = MlConfigVersion.CURRENT;
+        MlConfigVersion maxMlConfigVersion = MlConfigVersion.MINIMUM_COMPATIBLE;
+        for (DiscoveryNode node : nodes) {
+            String mlConfigVersionStr = node.getAttributes().get(mlConfigVersionNodeAttrName);
+            MlConfigVersion mlConfigVersion = fromString(mlConfigVersionStr);
+            if (mlConfigVersion.before(minMlConfigVersion)) {
+                minMlConfigVersion = mlConfigVersion;
+            }
+            if (mlConfigVersion.after(maxMlConfigVersion)) {
+                maxMlConfigVersion = mlConfigVersion;
+            }
+        }
+        return new Tuple<>(minMlConfigVersion, maxMlConfigVersion);
+    }
+
+    public static MlConfigVersion fromString(String str) {
```

Comment:
I don't know how feasible it is to get a negative version but should we use `Integer.parseUInt` or check to ensure that the resulting number isn't negative?

Reply:
Thanks Jonathan. I think the main issue here is that `MlConfigVersion` needs to be backwardly compatible with `Version` and hence it makes more sense to palm off the string parsing to `Version.fromString`